tensorboard/plugins/debugger/debugger_server_lib.py

```diff
@@ -0,0 +1,321 @@
+        the core runtime of a debugged Session::Run() call, after graph
+        partition. Therefore it may differ from the GraphDef available to
+        the general TensorBoard. For example, the GraphDef in general
+        TensorBoard may get partitioned for multiple devices (CPUs and GPUs),
+        each of which will generate a GraphDef event proto sent to this
+        method.
+      device_name: Name of the device on which the graph was created.
+      wall_time: An epoch timestamp (in microseconds) for the graph.
+    """
+    # For now, we do nothing with the graph def. However, we must define this
+    # method to satisfy the handler's interface. Furthermore, we may use the
+    # graph in the future (for instance to provide a graph if there is no graph
+    # provided otherwise).
+    del device_name
+    del wall_time
+    del graph_def
+
+  def on_value_event(self, event):
+    """Records the summary values based on an updated message from the debugger.
+
+    Logs an error message if writing the event to disk fails.
+
+    Args:
+      event: The Event proto to be processed.
+    """
+    if not event.summary.value:
+      tf.logging.warn("The summary of the event lacks a value.")
+      return
+
+    # The node name property is actually a watch key, which is a concatenation
+    # of several pieces of data.
+    watch_key = event.summary.value[0].node_name
+    if not watch_key.endswith(constants.DEBUG_NUMERIC_SUMMARY_SUFFIX):
+      # Ignore events that lack a DebugNumericSummary.
+      # NOTE(chizeng): We may later handle other types of debug ops.
+      return
+
+    # We remove the constants.DEBUG_NUMERIC_SUMMARY_SUFFIX from the end of the
+    # watch name because it is not distinguishing: every health pill entry ends
+    # with it.
+    node_name_and_output_slot = watch_key.rstrip(
+        constants.DEBUG_NUMERIC_SUMMARY_SUFFIX)
+
+    tensor = event.summary.value[0].tensor
+    dim = tensor.tensor_shape.dim
+    if (len(dim) != 1 or
+        dim[0].size < constants.MIN_DEBUG_NUMERIC_SUMMARY_TENSOR_LENGTH):
+      tf.logging.warning("Health-pill tensor either lacks a dimension or is "
+                         "size incorrectly. len(dim): %d, size of dim: %d",
+                         len(dim), dim[0].size)
+      return
+
+    match = re.match(r"^(.*):(\d+)$", node_name_and_output_slot)
+    if not match:
+      tf.logging.warning(
+          ("A event with a health pill has an invalid node name and output "
+           "slot combination, (i.e., an unexpected debug op): %r"),
+          node_name_and_output_slot)
+      return None
+
+    if self._session_run_index >= 0:
+      event.step = self._session_run_index
+    else:
+      # Data from parameter servers (or any graphs without a master) do not
+      # contain core metadata. So the session run count is missing. Set its
+      # value to a microsecond epoch timestamp.
+      event.step = int(time.time() * 1e6)
```

Comment:
Cool—as long as you say it's okay, I'm happy with it :-). Thanks for the explanation.